File: programcreek/python/01/01-10-3sum-closest.py
import sys

def threeSumClosest(nums, target) :
    nums.sort()
    minValue = sys.maxint
    result = 0


    for i in range(len(nums)):
        j = i + 1
        k = len(nums) - 1

        while j < k:

            total = nums[i] + nums[j] + nums[k]
            diff = abs(target - total)

            if diff == 0:
                return total

            if diff < minValue:
                minValue = diff
                result = total

            if total < target:
                j += 1
            else :
                k -= 1

    return result



print threeSumClosest([-1, 2, 1, -4], 1);

File: programcreek/js/01/01-10-3sum-closest.js


function threeSumClosest(nums, target) {

    nums.sort();
    var result = 0;
    var bestAns = Number.MAX_VALUE;

    for ( var i = 0 ; i < nums.length; i ++ ) {
        var j = i + 1;
        var k = nums.length - 1;

        while (j < k) {

            var total = nums[i] + nums[j] + nums[k];
            var diff = Math.abs(total - target);

            if (diff == 0) return total;

            if (diff < bestAns) {
                bestAns = diff;
                result = total;
            }

            if (total < target) {
                j ++;
            }
            else {
                k --;
            }

        }
    }

    return result;

}



console.log(threeSumClosest([-1, 2, 1, -4], 1));